clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2473,9 +2475,116 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
 
 } // namespace
 
+class SYCLKernelNameTypeVisitor
+    : public TypeVisitor<SYCLKernelNameTypeVisitor>,
+      public ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor> {
+  Sema &S;
+  SourceLocation Loc;
+  using InnerTypeVisitor = TypeVisitor<SYCLKernelNameTypeVisitor>;
+  using InnerTAVisitor =
+      ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor>;
+
+public:
+  SYCLKernelNameTypeVisitor(Sema &S, SourceLocation Loc) : S(S), Loc(Loc) {}
+
+  void Visit(QualType T) {
+    if (T.isNull())
+      return;
+    const CXXRecordDecl *RD = T->getAsCXXRecordDecl();
+    if (!RD)
+      return;
+    // If KernelNameType has template args visit each template arg via
+    // ConstTemplateArgumentVisitor
+    if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
+      const TemplateArgumentList &Args = TSD->getTemplateArgs();
+      for (unsigned I = 0; I < Args.size(); I++) {
+        const TemplateArgument &TemplateArg = Args[I];
+        Visit(TemplateArg);
+      }
+    } else {
+      InnerTypeVisitor::Visit(T.getTypePtr());
+    }
+  }
+
+  void Visit(const TemplateArgument &TA) { InnerTAVisitor::Visit(TA); }
+
+  void VisitEnumType(const EnumType *T) {
+    const EnumDecl *ED = T->getDecl();
+    if (!ED->isScoped() && !ED->isFixed()) {
+      S.Diag(Loc, diag::err_sycl_kernel_incorrectly_named) << 2;
+      S.Diag(ED->getSourceRange().getBegin(), diag::note_entity_declared_at)
+          << ED;
+      return;
+    }
+  }
+
+  void VisitRecordType(const RecordType *T) {
+    return VisitTagDecl(T->getDecl());
+  }
+
+  void VisitTagDecl(const TagDecl *Tag) {
+    bool UnnamedLambdaEnabled =
+        S.getASTContext().getLangOpts().SYCLUnnamedLambda;
+    if (Tag && !Tag->getDeclContext()->isTranslationUnit() &&
```

Comment:
Is it possible that `Tag` is nullptr here?
Will we enter `VisitTagDecl` function in this case?

Reply:
There is a null check here the first time we call Visit() [L2492](https://github.com/intel/llvm/pull/2474/files#diff-c88a590aca6d5f8e92dd1f8b6ec05a49R2492)
I added another check for the TemplateArgs here : [L2511](https://github.com/intel/llvm/pull/2474/files#diff-c88a590aca6d5f8e92dd1f8b6ec05a49R2511), so Tag should not be null here.
We will not reach `VisitTagDecl()` if any of the types were null.
I can remove the `if(Tag)` check here if you prefer.

Reply:
> I can remove the if(Tag) check here if you prefer.

Yes, please.